pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java

```diff
@@ -100,26 +111,58 @@ public class QueryEnvironment {
   private final FrameworkConfig _config;
   private final CalciteCatalogReader _catalogReader;
   private final HepProgram _optProgram;
-  private final HepProgram _traitProgram;
-
-  // Pinot extensions
-  private final TableCache _tableCache;
-  private final WorkerManager _workerManager;
+  private final Config _envConfig;
 
-  public QueryEnvironment(String database, TableCache tableCache, @Nullable WorkerManager workerManager) {
-    PinotCatalog catalog = new PinotCatalog(tableCache, database);
+  public QueryEnvironment(Config config) {
+    _envConfig = config;
+    String database = config.getDatabase();
+    PinotCatalog catalog = new PinotCatalog(config.getTableCache(), database);
     CalciteSchema rootSchema = CalciteSchema.createRootSchema(false, false, database, catalog);
     _config = Frameworks.newConfigBuilder().traitDefs().operatorTable(PinotOperatorTable.instance())
         .defaultSchema(rootSchema.plus()).sqlToRelConverterConfig(PinotRuleUtils.PINOT_SQL_TO_REL_CONFIG).build();
     _catalogReader = new CalciteCatalogReader(rootSchema, List.of(database), _typeFactory, CONNECTION_CONFIG);
     _optProgram = getOptProgram();
-    _traitProgram = getTraitProgram();
-    _tableCache = tableCache;
-    _workerManager = workerManager;
   }
 
-  private PlannerContext getPlannerContext() {
-    return new PlannerContext(_config, _catalogReader, _typeFactory, _optProgram, _traitProgram);
+  public QueryEnvironment(String database, TableCache tableCache, @Nullable WorkerManager workerManager) {
+    this(configBuilder()
+        .database(database)
+        .tableCache(tableCache)
+        .workerManager(workerManager)
+        .build());
+  }
+
+  /**
+   * Returns a planner context that can be used to either parse, explain or execute a query.
+   */
+  private PlannerContext getPlannerContext(SqlNodeAndOptions sqlNodeAndOptions) {
+    boolean useImplicitColocated;
+    PinotImplicitTableHintRule.PartitionTableFinder ptf;
+    String useImplicitColocatedOptionValue = sqlNodeAndOptions.getOptions()
```

Comment:
We want to allow user to explicitly turn this off by setting a query option. Here we want to check `null` to differentiate with explicit `false`

Reply:
I've rewritten this logic. Please review the new code. The idea is that:
- If the option is `true`, we force try to use the optimization, failing if workerManager is not defined.
- If the option is `false`, we do not use the optimization.
- If the option is not defined (`null`), we use the default (decided by PinotConfig)
- In any other case, we fail saying it is not a valid value.

Reply:
The logic looks correct, but IMO not very readable.. In other places we usually handle it with the following logic flow (basically make the override very explicit):
```
boolean useImplicitColocated = _envConfig.useImplicitColocatedByDefault();
if (useImplicitColocatedOptionValue != null) {
  useImplicitColocated = Boolean.parseBoolean(useImplicitColocatedOptionValue);
}
if (useImplicitColocated) {
  ...
} else {
  ...
}
```

Reply:
That is not exactly the same. Specifically it doesn't fail when providing a value that is not `true` or `false` but instead assumes anything different than `true` is false. But you are right, that code is not easy to read.

I've rewritten it to make it easier to read and keep the logic above.